addons/stock/models/stock_quant.py

```diff
@@ -170,40 +170,43 @@ def _search_on_hand(self, operator, value):
-                'inventory_quantity_auto_apply', False) or 0
-            # Create an empty quant or write on a similar one.
-            product = self.env['product.product'].browse(vals['product_id'])
-            location = self.env['stock.location'].browse(vals['location_id'])
-            lot_id = self.env['stock.production.lot'].browse(vals.get('lot_id'))
-            package_id = self.env['stock.quant.package'].browse(vals.get('package_id'))
-            owner_id = self.env['res.partner'].browse(vals.get('owner_id'))
-            quant = self._gather(product, location, lot_id=lot_id, package_id=package_id, owner_id=owner_id, strict=True)
-
-            if quant:
-                quant = quant[0].sudo()
+        quants = self.env['stock.quant']
+        for vals in vals_list:
+            if self._is_inventory_mode() and any(f in vals for f in ['inventory_quantity', 'inventory_quantity_auto_apply']):
+                allowed_fields = self._get_inventory_fields_create()
+                if any(field for field in vals.keys() if field not in allowed_fields):
+                    raise UserError(_("Quant's creation is restricted, you can't do this operation."))
+
+                inventory_quantity = vals.pop('inventory_quantity', False) or vals.pop(
+                    'inventory_quantity_auto_apply', False) or 0
+                # Create an empty quant or write on a similar one.
+                product = self.env['product.product'].browse(vals['product_id'])
+                location = self.env['stock.location'].browse(vals['location_id'])
+                lot_id = self.env['stock.production.lot'].browse(vals.get('lot_id'))
+                package_id = self.env['stock.quant.package'].browse(vals.get('package_id'))
+                owner_id = self.env['res.partner'].browse(vals.get('owner_id'))
+                quant = self._gather(product, location, lot_id=lot_id, package_id=package_id, owner_id=owner_id, strict=True)
+
+                if quant:
+                    quant = quant[0].sudo()
+                else:
+                    quant = self.sudo().create(vals)
+                # Set the `inventory_quantity` field to create the necessary move.
+                quant.inventory_quantity = inventory_quantity
+                quant.user_id = vals.get('user_id', self.env.user.id)
+                quant.inventory_date = fields.Date.today()
+                quants |= quant
             else:
-                quant = self.sudo().create(vals)
-            # Set the `inventory_quantity` field to create the necessary move.
-            quant.inventory_quantity = inventory_quantity
-            quant.user_id = vals.get('user_id', self.env.user.id)
-            quant.inventory_date = fields.Date.today()
-
-            return quant
-        res = super(StockQuant, self).create(vals)
-        if self._is_inventory_mode():
-            res._check_company()
-        return res
+                quant = super().create(vals)
```

Comment:
you could recreate a new list with vals containing `inventory_quantity` and `inventory_quantity_auto_apply` in order to call the super to create only once